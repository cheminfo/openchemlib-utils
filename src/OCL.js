let OCL;

export function getOCL() {
  if (!OCL) {
    throw new Error('OCL has to be initialized using initOCL(OCL)');
  }
  return OCL;
}
/**
 * 
 * @param {OCL} newOCL 
 * @param {object} [options={}]
 * @param {boolean} [options.keepExisting=false] Will keep current OCL if we initialize more than once
 */
export function initOCL(newOCL, options = {}) {
  const { keepExisting = false } = options;
  if (OCL && keepExisting) return;
  if (OCL && OCL !== newOCL) {
    throw new Error('OCL-utils was already initialized with a different OCL');
  }
  OCL = newOCL;
}
