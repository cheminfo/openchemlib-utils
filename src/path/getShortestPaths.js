/**
 * Get the shortest path between each pair of atoms in the molecule
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.maxLength=4]
 * @returns {Array<Array>} A matrix containing on each cell (i,j) the shortest path from atom i to atom j
 */
export function getShortestPaths(molecule, options = {}) {
  const OCL = molecule.getOCL();
  const { fromLabel = '', toLabel = '', maxLength = 3 } = options;

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  const nbAtoms = molecule.getAllAtoms();
  let allShortestPaths = new Array(nbAtoms);
  for (let i = 0; i < nbAtoms; i++) {
    allShortestPaths[i] = new Array(nbAtoms);
  }

  for (let from = 0; from < nbAtoms; from++) {
    allShortestPaths[from][from] = [from];
    for (let to = from + 1; to < nbAtoms; to++) {
      if (
        (fromAtomicNumber === 0 ||
          molecule.getAtomicNo(from) === fromAtomicNumber) &&
        (toAtomicNumber === 0 || molecule.getAtomicNo(to) === toAtomicNumber)
      ) {
        let path = [];
        molecule.getPath(path, from, to, maxLength);
        if (path.length) {
          allShortestPaths[from][to] = path.slice();
          allShortestPaths[to][from] = path.reverse();
        } else {
          allShortestPaths[from][to] = null;
          allShortestPaths[to][from] = null;
        }
      } else {
        allShortestPaths[from][to] = null;
        allShortestPaths[to][from] = null;
      }
    }
  }

  return allShortestPaths;
}
