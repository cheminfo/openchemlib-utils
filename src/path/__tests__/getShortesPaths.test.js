import OCL from 'openchemlib';
import { expect, it, describe } from 'vitest';

import { getShortestPaths } from '../getShortestPaths';

describe('getShortestPaths', () => {
  it('CC', () => {
    const molecule = OCL.Molecule.fromSmiles('CC');
    molecule.addImplicitHydrogens();
    const paths = getShortestPaths(molecule);
    //Atoms 0 and 1 are Carbons. So we must get path[0][1] = [0, 1]
    expect(paths[0][1]).toStrictEqual([0, 1]);
    //hydrogens 2, 3, and 4 must be attached to atom 0 and hydrogens 5, 6, and 7 must be attached to atom 1
    expect(paths[0][2]).toStrictEqual([0, 2]);
    expect(paths[0][3]).toStrictEqual([0, 3]);
    expect(paths[0][4]).toStrictEqual([0, 4]);
    expect(paths[1][5]).toStrictEqual([1, 5]);
    expect(paths[1][6]).toStrictEqual([1, 6]);
    expect(paths[1][7]).toStrictEqual([1, 7]);
    //hydrogens 2 and 3 must be attached to same carbon. The path must be symetric
    expect(paths[2][3]).toStrictEqual([2, 0, 3]);
    expect(paths[3][2]).toStrictEqual([3, 0, 2]);
    //hydrogens 5 and 7 must be attached to same carbon. The path must be symetric
    expect(paths[5][7]).toStrictEqual([5, 1, 7]);
    expect(paths[7][5]).toStrictEqual([7, 1, 5]);
    //hydrogens 3 and 6 must be attached to different carbons, so the path must be longest possible.
    expect(paths[3][6]).toStrictEqual([3, 0, 1, 6]);
    expect(paths[6][3]).toStrictEqual([6, 1, 0, 3]);
  });

  it('C.C', () => {
    const molecule = OCL.Molecule.fromSmiles('C.C');

    molecule.addImplicitHydrogens();
    const paths = getShortestPaths(molecule);
    expect(paths[0]).toStrictEqual([
      [0],
      null,
      [0, 2],
      [0, 3],
      [0, 4],
      [0, 5],
      null,
      null,
      null,
      null,
    ]);
    expect(paths).toMatchSnapshot();
  });
});
