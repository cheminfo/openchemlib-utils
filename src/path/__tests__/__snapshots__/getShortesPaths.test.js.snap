// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getShortestPaths C.C 1`] = `
[
  [
    [
      0,
    ],
    null,
    [
      0,
      2,
    ],
    [
      0,
      3,
    ],
    [
      0,
      4,
    ],
    [
      0,
      5,
    ],
    null,
    null,
    null,
    null,
  ],
  [
    null,
    [
      1,
    ],
    null,
    null,
    null,
    null,
    [
      1,
      6,
    ],
    [
      1,
      7,
    ],
    [
      1,
      8,
    ],
    [
      1,
      9,
    ],
  ],
  [
    [
      2,
      0,
    ],
    null,
    [
      2,
    ],
    [
      2,
      0,
      3,
    ],
    [
      2,
      0,
      4,
    ],
    [
      2,
      0,
      5,
    ],
    null,
    null,
    null,
    null,
  ],
  [
    [
      3,
      0,
    ],
    null,
    [
      3,
      0,
      2,
    ],
    [
      3,
    ],
    [
      3,
      0,
      4,
    ],
    [
      3,
      0,
      5,
    ],
    null,
    null,
    null,
    null,
  ],
  [
    [
      4,
      0,
    ],
    null,
    [
      4,
      0,
      2,
    ],
    [
      4,
      0,
      3,
    ],
    [
      4,
    ],
    [
      4,
      0,
      5,
    ],
    null,
    null,
    null,
    null,
  ],
  [
    [
      5,
      0,
    ],
    null,
    [
      5,
      0,
      2,
    ],
    [
      5,
      0,
      3,
    ],
    [
      5,
      0,
      4,
    ],
    [
      5,
    ],
    null,
    null,
    null,
    null,
  ],
  [
    null,
    [
      6,
      1,
    ],
    null,
    null,
    null,
    null,
    [
      6,
    ],
    [
      6,
      1,
      7,
    ],
    [
      6,
      1,
      8,
    ],
    [
      6,
      1,
      9,
    ],
  ],
  [
    null,
    [
      7,
      1,
    ],
    null,
    null,
    null,
    null,
    [
      7,
      1,
      6,
    ],
    [
      7,
    ],
    [
      7,
      1,
      8,
    ],
    [
      7,
      1,
      9,
    ],
  ],
  [
    null,
    [
      8,
      1,
    ],
    null,
    null,
    null,
    null,
    [
      8,
      1,
      6,
    ],
    [
      8,
      1,
      7,
    ],
    [
      8,
    ],
    [
      8,
      1,
      9,
    ],
  ],
  [
    null,
    [
      9,
      1,
    ],
    null,
    null,
    null,
    null,
    [
      9,
      1,
      6,
    ],
    [
      9,
      1,
      7,
    ],
    [
      9,
      1,
      8,
    ],
    [
      9,
    ],
  ],
]
`;
